server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -82,51 +93,128 @@ public void onNext(final StreamRequest request) {
                 requireObjects.add(object);
+                runOrEnqueue(requireObjects, () -> {
+                    Object[] objs = referenceObjects.stream().map(ExportObject::get).toArray();
+                    messageStream.onData(data.getPayload().asReadOnlyByteBuffer(), objs);
+                });
+            }
+        }
+
+        private final Queue<StreamOperation> operations = new ConcurrentLinkedQueue<>();
+        private final AtomicBoolean running = new AtomicBoolean(false);
+
+        // These methods are intended to roughly behave like SerializingExecutor(directExecutor()) in that only one can
+        // be running at a time, and will be started on the current thread, with the distinction that submitted work
+        // will continue off-thread and will signal when it is finished
+        private void runOrEnqueue(StreamOperation runnable) {
+            runOrEnqueue(Collections.emptyList(), runnable);
+        }
+
+        private void runOrEnqueue(Collection<? extends ExportObject<?>> additExports, StreamOperation operation) {
+            List<ExportObject<?>> requirements = new ArrayList<>(additExports);
+            requirements.add(object);
+
+            StreamOperation runnable = () -> {
                 session.nonExport()
-                        .require(requireObjects)
-                        .onError(responseObserver)
+                        .onErrorHandler(this::onError)
+                        .require(requirements)
                         .submit(() -> {
-                            Object[] objs = referenceObjects.stream().map(ExportObject::get).toArray();
-                            // TODO: Should we try/catch this to recover/not close from plugin errors? Or make the
-                            // client
-                            // recover? Would be useful if the plugin is complex
-                            messageStream.onData(data.getPayload().asReadOnlyByteBuffer(), objs);
+                            // Run the specified work. Note that we're not concerned about exceptions, the stream will
+                            // be dead (via onError) and won't be used again.
+                            try {
+                                operation.run();
+                            } catch (ObjectType.ObjectCommunicationException e) {
+                                throw new IllegalStateException("Error occurred communicating with client", e);
+                            }
+
+                            // Set running to false (it must be true at this time) so that any new work can race being
+                            // added
+                            running.set(false);
+                            doWork();
                         });
+            };
+            // gRPC guarantees we can't race in this method
+            operations.add(runnable);
+            if (running.compareAndSet(false, true)) {
```

Comment:
I'm not sure what this guard is accomplishing... isn't the same exact guard already in doWork? In fact, isn't this almost guaranteeing we break our state? (ie, we'll set it to true, but won't ever set it to false?)

Reply:
You're right, this doesn't work in the case where the newly added item wins the race... which means it will never work right (only in the case where it successfully _starts_ doing work once). I'll take another stab at it.